connector/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/AlgorithmRegisty.scala

```diff
@@ -0,0 +1,157 @@
+
+  def loadModel(path: String): Model[_]
+
+  def loadEstimator(path: String): Estimator[_]
+
+  protected def getEstimatorWriter(estimator: Estimator[_]): MLWriter
+
+  protected def getModelWriter(model: Model[_]): MLWriter
+
+  def _save(
+      writer: MLWriter,
+      path: String,
+      overwrite: Boolean,
+      options: Map[String, String]): Unit = {
+    if (overwrite) {
+      writer.overwrite()
+    }
+    options.map { case (k, v) => writer.option(k, v) }
+    writer.save(path)
+  }
+
+  def saveModel(
+      model: Model[_],
+      path: String,
+      overwrite: Boolean,
+      options: Map[String, String]): Unit = {
+    _save(getModelWriter(model), path, overwrite, options)
+  }
+
+  def saveEstimator(
+      estimator: Estimator[_],
+      path: String,
+      overwrite: Boolean,
+      options: Map[String, String]): Unit = {
+    _save(getEstimatorWriter(estimator), path, overwrite, options)
+  }
+}
+
+class LogisticRegressionAlgorithm extends Algorithm {
+
+  override def initiateEstimator(uid: String): Estimator[_] = {
+    new ml.classification.LogisticRegression(uid)
+  }
+
+  override def loadModel(path: String): Model[_] = {
+    ml.classification.LogisticRegressionModel.load(path)
+  }
+
+  override def loadEstimator(path: String): Estimator[_] = {
+    ml.classification.LogisticRegression.load(path)
+  }
+
+  protected override def getModelWriter(model: Model[_]): MLWriter = {
+    model.asInstanceOf[ml.classification.LogisticRegressionModel].write
+  }
+
+  protected override def getEstimatorWriter(estimator: Estimator[_]): MLWriter = {
+    estimator.asInstanceOf[ml.classification.LogisticRegression].write
+  }
+
+  override def getModelAttr(
+      model: Model[_],
+      name: String): Either[proto.MlCommandResponse, DataFrame] = {
+    val lorModel = model.asInstanceOf[ml.classification.LogisticRegressionModel]
+    // TODO: hasSummary
+    name match {
+      case "hasSummary" => Left(Serializer.serialize(lorModel.hasSummary))
+      case "numClasses" => Left(Serializer.serialize(lorModel.numClasses))
+      case "numFeatures" => Left(Serializer.serialize(lorModel.numFeatures))
+      case "intercept" => Left(Serializer.serialize(lorModel.intercept))
+      case "interceptVector" => Left(Serializer.serialize(lorModel.interceptVector))
+      case "coefficients" => Left(Serializer.serialize(lorModel.coefficients))
+      case "coefficientMatrix" => Left(Serializer.serialize(lorModel.coefficientMatrix))
+      case _ =>
+        throw new IllegalArgumentException()
+    }
+  }
+
+  override def getModelSummaryAttr(
+      model: Model[_],
+      name: String,
+      datasetOpt: Option[DataFrame]): Either[proto.MlCommandResponse, DataFrame] = {
+    val lorModel = model.asInstanceOf[ml.classification.LogisticRegressionModel]
+    val summary = if (datasetOpt.isDefined) {
+      lorModel.evaluate(datasetOpt.get)
+    } else {
+      lorModel.summary
+    }
```

Comment:
```suggestion
    val summary = datasetOpt match {
      case Some(dataset) => lorModel.evaluate(dataset)
      case None => lorModelSummary
    }
```

Can we use pattern maching here?